docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -253,14 +255,240 @@ performance would be poor for services with many nodes.
+  throws (1:InvalidRequestException ire, 2:UnavailableException ue, 3:TimedOutException te)
+```
+
+Implementing this endpoint on the Cassandra side was not too difficult. It may seem a little wasteful in that this may
+require keys and predicates to be specified more than once, but for transactions2 these would likely be mostly
+distinct.
+
+While this improved performance, we still faced significant regressions relative to the v1 cell loader. We determined
+that this was because Cassandra has a worker pool for loading values to satisfy a ``ReadCommand``, but the calling
+thread in requests is also allowed to participate. Thus, creating large batches would turn out to likely be detrimental
+to read performance, even when the Cassandra nodes are actually able to handle higher concurrency safely.
+
+#### Selective Batching
+
+We thus settled on a compromise between sending large singular requests and inundating Cassandra with smaller ones;
+unlike in the original ``CellLoader``, we make the batch parameters configurable. There are two parameters:
+
+- cross column load batch limit (CC); we may combine requests for different columns in one call to the DB, but merged
+  calls will not exceed this size.
+- single query load batch limit (SQ); a single request should never be larger than this size. We expect SQ >= CC.
+
+We still partition requests to load cells by column first. Thereafter,
+
+- if for a given column the number of cells is at least CC, then the cells for that column will exclusively take up
+  one or more batches, with no batch having size greater than SQ.
+- otherwise, the cells may be combined with cells in other columns, in batches of size up to CC. There is no guarantee
+  that all cells for a given column will be in the same batch.
+
+In terms of implementation, we simply maintain a list of cells eligible for cross-column batching, and partition this
+list into contiguous groups of size CC. In this way, no row key will be included more than twice. It may be possible
+to reduce the amount of data sent over the wire to Cassandra and possibly some of the internal read bandwidth by
+solving the underlying bin-packing problem to ensure that each row-key only occurs once; consider that we duplicate
+many keys if we want to load CC - 1 cells from many columns. This may be worth considering in the future (while
+bin-packing is NP-complete, an algorithm like first-fit decreasing will give us a good approximation), but we have not
+implemented it yet as the overhead is only a constant factor, and in many cases with transactions2 we expect the
+number of cells per column to be small. Consider that assuming a uniform distribution, even if a single transaction 
+reads 1,000,000 values with PQ / NP = 312,500, the maximum batch size will probably not exceed 20.
+
+#### Benchmarking
+
+We tested the selective batching cell loader against the original algorithm ("CL1") and a full-batching algorithm that
```

Comment:
Tiny bit confused, likely due to me misunderstanding the space, are the different experiments tx1 workloads, general workloads, tx2 workloads?

are you here also evaluating whether or not CL2 could be *the* `CellLoader` or is this solely for tx2? 

can we also get a chart? haha less serious about this one though

Reply:
1. We didn't actually test a tx1 workload (I think by this point we were quite certain that we wanted to go ahead with tx2). The 100/100 static and 1000/10 static are general workloads that users might have for (which is important because CL2 is used for all AtlasDB gets, not just the ones from the transactions table).

2. Yeah, we are - I've now clarified this in the text. CL2 is _the_ `CellLoader` for general workflows now, actually.

3. That's reasonable, will add one soon (though not in the commit I'm going to push right after this!)